File: typescript/src/sso.ts
/**
 * SSO administration types — connections, JIT rules, events, and enforcement
 * readiness. These mirror the wire shapes served by the v1-sso edge function.
 *
 * Usage:
 *
 * ```ts
 * import type { SsoConnection, SsoReadiness } from "@atlasent/sdk";
 *
 * const res = await fetch("/functions/v1/v1-sso/connections", { ... });
 * const { connections } = await res.json() as { connections: SsoConnectionWire[] };
 * const typed = connections.map(wireToSsoConnection);
 * ```
 */

// ── Connections ───────────────────────────────────────────────────────────────

/** An SSO connection record (SAML 2.0 or OIDC). */
export interface SsoConnection {
  id: string;
  organizationId: string;
  name: string;
  protocol: "saml" | "oidc";
  idpEntityId: string;
  metadataUrl: string | null;
  metadataXml: string | null;
  emailDomain: string | null;
  enforceForDomain: boolean;
  isActive: boolean;
  supabaseProviderId: string | null;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

/** Wire (snake_case) shape for SSO connection responses. */
export interface SsoConnectionWire {
  id: string;
  organization_id: string;
  name: string;
  protocol: "saml" | "oidc";
  idp_entity_id: string;
  metadata_url: string | null;
  metadata_xml: string | null;
  email_domain: string | null;
  enforce_for_domain: boolean;
  is_active: boolean;
  supabase_provider_id: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export function wireToSsoConnection(w: SsoConnectionWire): SsoConnection {
  return {
    id: w.id,
    organizationId: w.organization_id,
    name: w.name,
    protocol: w.protocol,
    idpEntityId: w.idp_entity_id,
    metadataUrl: w.metadata_url,
    metadataXml: w.metadata_xml,
    emailDomain: w.email_domain,
    enforceForDomain: w.enforce_for_domain,
    isActive: w.is_active,
    supabaseProviderId: w.supabase_provider_id,
    createdBy: w.created_by,
    createdAt: w.created_at,
    updatedAt: w.updated_at,
  };
}

// ── JIT provisioning rules ───────────────────────────────────────────────────

export type SsoRole = "owner" | "admin" | "approver" | "member" | "viewer";

/** A JIT provisioning rule that maps an IdP claim to an org role. */
export interface SsoJitRule {
  id: string;
  connectionId: string;
  organizationId: string;
  claimAttribute: string;
  claimValue: string;
  grantedRole: SsoRole;
  precedence: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

/** Wire (snake_case) shape for JIT rule responses. */
export interface SsoJitRuleWire {
  id: string;
  connection_id: string;
  organization_id: string;
  claim_attribute: string;
  claim_value: string;
  granted_role: SsoRole;
  precedence: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export function wireToSsoJitRule(w: SsoJitRuleWire): SsoJitRule {
  return {
    id: w.id,
    connectionId: w.connection_id,
    organizationId: w.organization_id,
    claimAttribute: w.claim_attribute,
    claimValue: w.claim_value,
    grantedRole: w.granted_role,
    precedence: w.precedence,
    isActive: w.is_active,
    createdAt: w.created_at,
    updatedAt: w.updated_at,
  };
}

// ── Events ───────────────────────────────────────────────────────────────────

/**
 * An SSO lifecycle event — login, session, config change, break-glass, or
 * JIT provisioning.
 */
export interface SsoEvent {
  id: string;
  organizationId: string;
  connectionId: string | null;
  eventType: string;
  actorEmail: string | null;
  payload: Record<string, unknown>;
  occurredAt: string;
}

/** Wire (snake_case) shape for SSO event responses. */
export interface SsoEventWire {
  id: string;
  organization_id: string;
  connection_id: string | null;
  event_type: string;
  actor_email: string | null;
  payload: Record<string, unknown>;
  occurred_at: string;
}

export function wireToSsoEvent(w: SsoEventWire): SsoEvent {
  return {
    id: w.id,
    organizationId: w.organization_id,
    connectionId: w.connection_id,
    eventType: w.event_type,
    actorEmail: w.actor_email,
    payload: w.payload,
    occurredAt: w.occurred_at,
  };
}

// ── Enforcement state machine ─────────────────────────────────────────────────

/** Action to pass to `POST /v1/sso/enforce`. */
export type SsoEnforceAction = "enable" | "enforce";

/**
 * Four-boolean readiness checklist returned by `GET /v1/sso/status`.
 * All four must be `true` before enforcement is safe to activate.
 */
export interface SsoReadiness {
  /** At least one SSO connection row exists for the org. */
  connectionConfigured: boolean;
  /** At least one connection has been activated (registered with the IdP). */
  connectionTested: boolean;
  /** Break-glass access has been configured (non-default settings). */
  breakGlassSet: boolean;
  /** No unreviewed active service API keys exist. */
  serviceApiKeysReviewed: boolean;
}

/** Wire (snake_case) shape for the readiness response. */
export interface SsoReadinessWire {
  connection_configured: boolean;
  connection_tested: boolean;
  break_glass_set: boolean;
  service_api_keys_reviewed: boolean;
}

export function wireToSsoReadiness(w: SsoReadinessWire): SsoReadiness {
  return {
    connectionConfigured: w.connection_configured,
    connectionTested: w.connection_tested,
    breakGlassSet: w.break_glass_set,
    serviceApiKeysReviewed: w.service_api_keys_reviewed,
  };
}

File: python/atlasent/sso.py
"""Pydantic models for the ``/v1/sso/*`` wire surface.

Source of truth: ``atlasent-api/supabase/functions/v1-sso/handler.ts``
plus the table schemas in
``atlasent-api/supabase/migrations/0041_sso_connections.sql``.

Mirrors :mod:`atlasent.sso` in the TypeScript SDK. Field names are
snake_case because that is what the server emits — the SSO admin
surface returns persisted DB rows verbatim and the SDK does not
translate them into a different convention.

⚠ Upstream sync TODO: these models should ultimately be sourced from
``atlasent-api/packages/types/src/index.ts`` (the canonical
``@atlasent/types`` package). They are mirrored here as a stop-gap
until the upstream PR lands; once the upstream symbols are published
this module should re-export from a single canonical location.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

SsoProtocol = Literal["saml", "oidc"]
"""Identity-provider protocol enum.

Matches the table CHECK ``protocol IN ('saml', 'oidc')`` in
migration 0041. Today the handler ships SAML; OIDC is the planned
second protocol.
"""

SsoCanonicalRole = Literal["owner", "admin", "approver", "member", "viewer"]
"""Roles a JIT rule can grant.

Aligned with ``org_members.role`` via migration 0042. Tracks the
handler's ``CANONICAL_ROLES`` constant (the table CHECK is wider —
``operator`` / ``auditor`` are present but the handler rejects them).
"""

SsoEventType = Literal[
    "login_success",
    "login_denied",
    "jit_provisioned",
    "role_changed",
    "connection_created",
    "connection_updated",
    "connection_deleted",
    "connection_activated",
    "connection_deactivated",
]
"""SSO event-type tags persisted into ``sso_events.event_type``.

Drawn from the table CHECK constraint in
``0041_sso_connections.sql``. The handler emits the connection
lifecycle plus ``login_denied``; the assertion-side tags
(``login_success``, ``jit_provisioned``, ``role_changed``) fire from
the Supabase post-assertion hook.
"""


class SsoConnection(BaseModel):
    """One row from ``sso_connections`` as returned by
    ``GET /v1/sso/connections`` and ``GET /v1/sso/connections/:id``.

    ``metadata_url`` and ``metadata_xml`` are mutually-permissive (the
    table CHECK requires at least one). ``enforce_for_domain`` is
    gated on the org's plan; see the handler's free-tier check.
    """

    id: str
    organization_id: str
    name: str
    protocol: SsoProtocol
    supabase_provider_id: str | None = None
    idp_entity_id: str
    metadata_url: str | None = None
    metadata_xml: str | None = None
    email_domain: str | None = None
    enforce_for_domain: bool = False
    is_active: bool = False
    created_at: str = ""
    updated_at: str = ""
    created_by: str | None = None

    model_config = {"extra": "allow"}


class SsoJitRule(BaseModel):
    """One row from ``sso_jit_provisioning_rules`` as returned by
    ``GET /v1/sso/jit-rules`` (optionally filtered by
    ``connection_id``).

    Matching semantics: the rule fires when
    ``claims[claim_attribute]`` contains ``claim_value`` (membership
    for array-shaped claims, equality for scalars). When multiple rules
    match the lowest ``precedence`` wins; ties broken by row id.
    """

    id: str
    connection_id: str
    organization_id: str
    claim_attribute: str
    claim_value: str
    granted_role: SsoCanonicalRole
    precedence: int = 100
    is_active: bool = True
    created_at: str = ""
    updated_at: str = ""

    model_config = {"extra": "allow"}


class SsoEvent(BaseModel):
    """One row from ``sso_events`` as returned by
    ``GET /v1/sso/events``.

    ``connection_id`` is nullable because a denial can fire before the
    inbound assertion is mapped to a connection. ``actor_email`` is
    nullable for system-actor events.
    """

    id: str
    organization_id: str
    connection_id: str | None = None
    event_type: SsoEventType
    actor_email: str | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    occurred_at: str = ""

    model_config = {"extra": "allow"}


class SsoEventsPage(BaseModel):
    """Response shape for ``GET /v1/sso/events``.

    ``next_cursor`` is ``None`` when the current page is the last —
    the handler emits ``null`` so callers can write a uniform
    pagination loop.
    """

    events: list[SsoEvent] = Field(default_factory=list)
    next_cursor: str | None = None

    model_config = {"extra": "allow"}
